### PR TITLE
Charlesmchen/honor attachment filenames

### PR DESCRIPTION
**Pull request by charlesmchen**
* Send filenames with outgoing attachments, if possible.
* Store incoming attachment filenames, if present.
* Write attachment data to disk using the attachment's filename, if possible.

PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
See: https://github.com/WhisperSystems/SignalServiceKit/pull/177

**Comment by michaelkirk** [comment]
Actually, this doesn't compile for me, but just missing a trivial change in NewGroupViewController

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 1]
```diff
@@ -2193,11 +2194,37 @@ - (void)resetFrame {
...
 {
-    OWSAssert([NSThread isMainThread]);
-    
     [UIUtil modalCompletionBlock]();
     [self resetFrame];
 
+    NSURL *referenceURL = [info valueForKey:UIImagePickerControllerReferenceURL];
+    if (!referenceURL) {
+        DDLogError(@"Could not retrieve reference URL for picked asset");
+        OWSAssert(0);
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: I hit this assertion failure on iOS9.3 (haven't tested on 10).

**steps:**

- Go to thread
- Paperclip
- Camera
- take a photo 
- use photo

**Comment by charlesmchen** [comment]
> Actually, this doesn't compile for me, but just missing a trivial change in NewGroupViewController

Fixed.

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: Fixed.

**Pull request merged**